test/test_circuitzoo_purification.jl

```diff
@@ -34,4 +43,86 @@ const bell = StabilizerState("XX ZZ")
             end
         end
     end
+
+    for rep in [QuantumOpticsRepr]
+        for leaveout in [:X, :Y, :Z]
+            r = Register(4, rep())
+            rnd = rand() / 4 + 0.5
+            noisy_pair = noisy_pair_func(rnd)
+            initialize!(r[1:2], noisy_pair)
+            initialize!(r[3:4], noisy_pair)
+            if Purify2to1(leaveout)(r[1:4]...)==true
+                @test abs(observable(r[1:2], projector(bell))) >= rnd
+            end
+        end
+    end
+end
+
+@testset "3to1" begin
+    for rep in [QuantumOpticsRepr, CliffordRepr]
+        for fixtwice in [:X, :Y, :Z]
+            # test that pure state gets mapped to pure state
+            r = Register(6, rep())
+            initialize!(r[1:6], bell⊗bell⊗bell)
+            @test Purify3to1(fixtwice)(r[1], r[2], [r[3], r[5]], [r[4], r[6]])==true
+            @test observable(r[1:2], projector(bell))≈1.0
+            # TODO: Should also taget qubits 1 and 2
+            for error in [:X, :Y, :Z], target in 3:6
+                r = Register(6, rep())
+                for i in 1:3
+                    initialize!(r[(2*i-1):(2*i)], bell)
+                end
+                apply!(r[target], Dict(:X=>X, :Y=>Y, :Z=>Z)[error])
+                @test Purify3to1(fixtwice)(r[1], r[2], [r[3], r[5]], [r[4], r[6]])==false
+            end
+            ## When [error, fixtwice] in {[X,Z], [Z,Y], [Y,X]} it yields true is that supposed to happen?
+            # for error in [:X, :Y, :Z], target in 1:2
+            #     r = Register(6, rep())
+            #     for i in 1:3
+            #         initialize!(r[(2*i-1):(2*i)], bell)
+            #     end
+            #     apply!(r[target], Dict(:X=>X, :Y=>Y, :Z=>Z)[error])
+            #     @test Purify3to1(fixtwice)(r[1], r[2], [r[3], r[5]], [r[4], r[6]])==false
+            # end
+        end
+    end
+    # testing fidelity - Error when using CliffordRepr
+    for rep in [QuantumOpticsRepr]
+        for fixtwice in [:X, :Y, :Z]
+            r = Register(6, rep())
+            rnd = rand() / 4 + 0.5
+            noisy_pair = noisy_pair_func(rnd)
+            initialize!(r[1:2], noisy_pair)
+            initialize!(r[3:4], noisy_pair)
+            initialize!(r[5:6], noisy_pair)
+            if Purify3to1(fixtwice)(r[1], r[2], [r[3], r[5]], [r[4], r[6]])==true
+                @test abs(observable(r[1:2], projector(bell))) >= rnd
+            end
+        end
+    end
+end
+
+@testset "Stringent" begin
+    for rep in [CliffordRepr, QuantumOpticsRepr]
+        r = Register(26, rep())
+        for i in 1:13
+            initialize!(r[(2*i-1):(2*i)], bell)
+        end
+        @test PurifyStringent()(r[1], r[2],
```

Comment:
When i do this the whole library gets kill signal because of the lack of ram.

Reply:
Oh, interesting, that is certainly a bug that should be fixed with the base library. Could you file an issue for it?

There is another quick fix: use the Clifford representation instead of state vectors. Kets and density matrices are exponentially expensive (as you have observed), but a Clifford tableaux simulation should be much cheaper and faster (even if it can not simulate all the types of dynamics that a ket-style simulation can)

Reply:
But there seems to be a problem when measuring the fidelity. It only works in QuantumOpticsRepr else it yields an error.